Disallow adding non-resolvable hostnames and hostnames that resolve to an "owned" IP in `cyhy-domain`

dav3r opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

This PR adds the following new features to the `cyhy-domain` script:
* If a hostname to be added to a CyHy entity does not resolve to at least one IPv4 address, it will not be added
* A new `--no-owned` option was added.  When used with `cyhy-domain add`, it prevents a hostname from being added if the hostname resolves to an IP address that is already "owned" in CyHy by an entity other than the default `CYHY` entity.
* If any domain-related errors occur with `cyhy-domain add`, a summary of those errors is output like this:
```
ERROR - foo.gov does not resolve to any IPv4 addresses
ERROR - bar.gov is already owned by: Entity1 (E1)
ERROR - baz.gov resolves to IP xxx.xxx.xxx.xxx owned by E2 in CyHy
ERROR - foobar.gov resolves to IP xxx.xxx.xxx.xxx owned by E3 in CyHy
ERROR - foobarbaz.gov is already owned by: Entity4 (E4)
ERROR - foobarbazoo.gov does not resolve to any IPv4 addresses

=== Summary of domain errors ===

Domains already owned by other CyHy entities:
bar.gov
foobarbaz.gov

Domains that do not resolve to any IPv4 addresses:
foo.gov
foobarbazoo.gov

Domains that resolve to IPs owned by other CyHy entities:
baz.gov
foobar.gov

EXITING without adding any domains.
```

Note: This PR was built on top of the changes in #121.  To view the differences from only this PR, click [here](https://github.com/cisagov/cyhy-core/compare/improvement/close-tix-for-removed-hostnames...improvement/disallow-nonresolvable-hostnames).

## 💭 Motivation and context ##

These additional features were requested by CyHy Ops team to:
* Ensure that only resolvable hostnames get added into CyHy inventory; this will reduce wasted time resolving non-resolvable hostnames every time the `cyhy-domainsync` ("housekeeping") script is run
* Have the ability to avoid adding any hostnames that resolve to IPs that are already "owned" in CyHy; this will help them with BOD compliance efforts

Resolves https://github.com/cisagov/cyhy-core/issues/119.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I tested the new code locally and verified that it operates as expected.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.
- [x] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [x] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Create a release (necessary if and only if the version was bumped).   This PR will be included in release `1.1.10`.
- [x] Deploy this change to Production.

dav3r committed:
Bump version of cyhy-domain from v1.1.0 to v1.2.0

dav3r committed:
Add a function to lookup IPv4 addresses in cyhy-domain

dav3r committed:
Update cyhy-domain to block the addition of hostnames that do not resolve

Also, output a summary if there are any errors when attempting to add domains.

dav3r committed:
Simplify print statements

dav3r committed:
Add --no-owned option to cyhy-domain

This option prevents the addition of domains that resolve to an IP owned (in CyHy) by a non-default entity (i.e. any entity other than the default owner "CYHY").

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -26,7 +28,9 @@ Notes:
 import sys
 
 # Third-Party Libraries
+from dns import resolver, exception
```
Small nitpick.

```suggestion
from dns import exception, resolver
```

mcdonnnj reviewed (approved):
I think this look alright. Nice work!

dav3r committed:
Sort dns imports alphabetically

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented on bin/cyhy-domain:
```diff
@@ -67,16 +71,76 @@ def get_existing_domains_and_owners(db):
+    except exception.DNSException as e:
+        print "Warning: Unexpected DNS exception for %s" % hostname
+        print str(e)
+    return set()
```
I think this is a little cleaner since it:
- Only includes what is absolutely necessary inside the `try` block
- Only has a single `return` statement

If you disagree then feel free to say so and I will resolve.

```suggestion
def get_ipv4_addresses(hostname):
    """Return a set of IPv4 addresses for the given hostname."""
    ip_set = set()
    try:
        answers = resolver.query(hostname, 'A')
    # Could not resolve domain name
    except resolver.NoNameservers as e:
        print "Warning: No nameservers available to resolve %s" % hostname
        print str(e)
    except resolver.Timeout as e:
        print "Warning: DNS query timed out for %s" % hostname
        print str(e)
    except exception.DNSException as e:
        print "Warning: Unexpected DNS exception for %s" % hostname
        print str(e)
    else:
        ip_list = [IPAddress(answer.address) for answer in answers]
        ip_set = set(ip_list)
    finally:
        return ip_set
```

dav3r replied:
Your way is 100% better, thanks!

jsf9k reviewed (approved):
Approved, but please consider my one comment.

dav3r committed:
Improve get_ipv4_addresses()

- Only include what is absolutely necessary inside the try block
- Only have a single return statement

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>
